src/test/java/com/google/jenkins/plugins/k8sengine/KubernetesEngineBuilderClusterTest.java

```diff
@@ -0,0 +1,294 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+
+import com.google.api.services.container.model.Cluster;
+import com.google.common.collect.ImmutableList;
+import com.google.jenkins.plugins.k8sengine.KubernetesEngineBuilder.DescriptorImpl;
+import com.google.jenkins.plugins.k8sengine.client.ClientFactory;
+import com.google.jenkins.plugins.k8sengine.client.ContainerClient;
+import hudson.AbortException;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import jenkins.model.Jenkins;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+/** Tests for the stapler methods for the cluster field of {@link KubernetesEngineBuilder}. */
+public class KubernetesEngineBuilderClusterTest {
+
+  private static final String TEST_PROJECT_ID = "test-project-id";
+  private static final String OTHER_PROJECT_ID = "other-project-id";
+  private static final String TEST_CLUSTER = "testCluster (us-west1-a)";
+  private static final String OTHER_CLUSTER = "otherCluster (us-east1-b)";
+  private static final String TEST_CREDENTIALS_ID = "test-credentials-id";
+
+  private static Jenkins jenkins;
+
+  @BeforeClass
+  public static void init() {
+    jenkins = Mockito.mock(Jenkins.class);
+  }
+
+  @Test
+  public void testDoFillClusterItemsEmptyWithEmptyCredentialsId() throws IOException {
+    DescriptorImpl descriptor = setUpClusterDescriptor(ImmutableList.of(), null, null);
+    ListBoxModel expected =
+        initExpected(ImmutableList.of(EMPTY_NAME), ImmutableList.of(EMPTY_VALUE));
+    ListBoxModel result = descriptor.doFillClusterItems(jenkins, null, null, TEST_PROJECT_ID);
+    assertNotNull(result);
+    assertListBoxModelEquals(expected, result);
+  }
+
+  @Test
+  public void testDoFillClusterItemsEmptyWithEmptyProjectId() throws IOException {
+    DescriptorImpl descriptor = setUpClusterDescriptor(ImmutableList.of(), null, null);
+    ListBoxModel expected =
+        initExpected(ImmutableList.of(EMPTY_NAME), ImmutableList.of(EMPTY_VALUE));
+    ListBoxModel result = descriptor.doFillClusterItems(jenkins, null, TEST_CREDENTIALS_ID, null);
+    assertNotNull(result);
+    assertListBoxModelEquals(expected, result);
+  }
+
+  @Test
+  public void testDoFillClusterItemsErrorMessageWithAbortException() throws IOException {
+    DescriptorImpl descriptor =
+        setUpClusterDescriptor(ImmutableList.of(), new AbortException(), null);
+    ListBoxModel expected =
+        initExpected(
+            ImmutableList.of(Messages.KubernetesEngineBuilder_CredentialAuthFailed()),
+            ImmutableList.of(EMPTY_VALUE));
+    ListBoxModel result =
+        descriptor.doFillClusterItems(jenkins, null, TEST_CREDENTIALS_ID, TEST_PROJECT_ID);
```

Comment:
Is null the value Stapler passes in when no cluster is currently selected?

Reply:
Good catch. Because this is a dropdown, no it won't ever be null, although because Strings.isNullOrEmpty is always used on this value it hasn't affected the validity of these tests. For projectId the same is true.